bazaarvoice/bv-ui-core
# SWAT-1245 -- Safer Promise polyfill (ponyfill)

[SWAT-1245](https://bits.bazaarvoice.com/jira/browse/SWAT-1245)

## Problem / Goal
Clients may sometimes polyfill Promise themselves with subpar implementations, which may cause our code that relies upon them to break.

## Solution
To account for cases where a client may have a preexisting version of a Promise polyfill that might not be spec-compliant, we're now polyfilling it ourselves if we detect anything other than native code. This also sneaks in some test code fixes, and explicitly imports our Promise polyfill wherever we're referencing Promise within the project. This should actually remove the need for polyfilling Promise in the scoutfile, which I've always hated that we had to do, since it's not using them directly.

## Verification
1. Load a browser that supports Promises natively (Chrome/FF/Safari).
2. Run the following in the JS console (this is the logic being tested against in the polyfill):
`typeof window.Promise === 'undefined' || !window.Promise.toString().match(/\[native code]/)`
3. Verify that you get a false output (this would result in the native Promise being returned).
4. Load a browser that doesn't support native Promises (IE 11).
5. Run the following in the JS console (this is the logic being tested against in the polyfill):
`typeof window.Promise === 'undefined' || !window.Promise.toString().match(/\[native code]/)`
6. Verify that you get a true output (this would result in the polyfilled Promise being returned).

## TODO
Write some actual test coverage around this. Since it's using a regex test for `[native code]`, we should be able to supply a stub function with that string in it, and verify that our polyfill either returns that function or another (the polyfilled version).
